### PR TITLE
py-opt-einsum: add v3.4.0

### PR DESCRIPTION
**Pull request by haampie**

### CODE REVIEW

File: var/spack/repos/builtin/packages/py-opt-einsum/package.py [hunk 1]
```diff
@@ -14,12 +14,21 @@ class PyOptEinsum(PythonPackage):
...
 
-    # https://github.com/dgasmith/opt_einsum/commit/7c8f193f90b6771a6b3065bb5cf6ec2747af8209
-    depends_on("python@:3.11", when="@:3.3", type=("build", "run"))
-    depends_on("py-setuptools", type="build")
+    with default_args(type=("build", "run")):
+        depends_on("python@3.8:", when="@3.4:")
+        # https://github.com/dgasmith/opt_einsum/commit/7c8f193f90b6771a6b3065bb5cf6ec2747af8209
+        depends_on("python@:3.11", when="@:3.3")
+
     depends_on("py-numpy@1.7:", type=("build", "run"))
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-opt-einsum/package.py]: No longer required?

**Review by adamjstewart — CHANGES_REQUESTED** [review]
Could also wrap more dependencies in `with default_args(type=...)` but not required

**Review by adamjstewart — DISMISSED** [review]

File: var/spack/repos/builtin/packages/py-jaxlib/package.py [hunk 2]
```diff
@@ -99,11 +99,12 @@ class PyJaxlib(PythonPackage, CudaPackage, ROCmPackage):
...
 
     with default_args(type=("build", "run")):
+        depends_on("python@3.10:", when="@0.4.31:")
+        depends_on("python@3.9:", when="@0.4.14:")
+        depends_on("python@3.8:", when="@0.4.6:")
         # Based on PyPI wheels
-        depends_on("python@3.10:3.12", when="@0.4.31:")
-        depends_on("python@3.9:3.12", when="@0.4.17:0.4.30")
-        depends_on("python@3.9:3.11", when="@0.4.14:0.4.16")
-        depends_on("python@3.8:3.11", when="@0.4.6:0.4.13")
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/py-jaxlib/package.py:106]: Can you add:
```python
depends_on("python@:3.13")
```
Someday we will have to add 3.14.

File: var/spack/repos/builtin/packages/py-jaxlib/package.py [hunk 3]
```diff
@@ -99,11 +99,12 @@ class PyJaxlib(PythonPackage, CudaPackage, ROCmPackage):
...
+        depends_on("python@:3.12", when="@:0.4.33")
```
Comment (haampie) [inline var/spack/repos/builtin/packages/py-jaxlib/package.py:107]:
```suggestion
        depends_on("python@:3.13")
        depends_on("python@:3.12", when="@:0.4.33")
```

File: var/spack/repos/builtin/packages/py-jaxlib/package.py (hunk 2 above)
Comment (haampie) [reply var/spack/repos/builtin/packages/py-jaxlib/package.py:106]: imo that leads to more work in the future, removing redundant bounds. I'd rather add upperbounds if things are known to break.

added this one though.

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**